core/src/main/scala/org/bitcoins/core/crypto/bip32/BIP32Path.scala

```diff
@@ -0,0 +1,70 @@
+package org.bitcoins.core.crypto.bip32
+
+import org.bitcoins.core.number.UInt32
+
+abstract class BIP32Path {
+  def children: Vector[BIP32Child]
+
+  override def toString: String =
+    children
+      .map {
+        case BIP32Child(index, hardened) =>
+          index.toString + (if (hardened) "'" else "")
+      }
+      .fold("m")((accum, curr) => accum + "/" + curr)
+
+}
+
+object BIP32Path {
+
+  /**
+    * The empty BIP32 path "m"
+    */
+  def empty: BIP32Path = BIP32PathImpl(Vector.empty)
+
+  private case class BIP32PathImpl(children: Vector[BIP32Child])
+      extends BIP32Path
+
+  def apply(children: Vector[BIP32Child]): BIP32Path = BIP32PathImpl(children)
+
+  def fromString(string: String): BIP32Path = {
+    val parts = string
+      .split("/")
+      .toVector
+      .map(_.trim)
+      .map(_.toLowerCase)
+
+    require(parts.nonEmpty,
+            """A BIP32 path string consists of "/"-separated elements""")
+
+    val head +: rest = parts
+    require(head == "m",
+            """The first element in a BIP32 path string must be "m"""")
+
+    val children = rest.map { str =>
+      val (index: String, hardened: Boolean) =
+        if (str.endsWith("'")) {
+          (str.dropRight(1), true)
+        } else {
+          (str, false)
+        }
+      BIP32Child(index.toInt, hardened)
+    }
+
+    BIP32PathImpl(children)
+  }
+}
+
+case class BIP32Child(index: Int, hardened: Boolean) {
+  require(index >= 0, "BIP32 child index must be positive!")
+
+  private val HARDENED_KEY_OFFSET = 2147483647 // pow(2, 31)
+
+  /**
+    * Converts this child to a BIP32 notation
+    * unsigned 32 bit integer
+    */
+  def toUInt32: UInt32 =
+    if (hardened) UInt32(HARDENED_KEY_OFFSET + index.toLong + 1)
```

Comment:
the `+1` scares me, where is that coming from ?

Reply:
Was because of wonky arithmetics, solved it by adding UInt32s directly